optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -77,13 +78,19 @@ public ProblemId_ getProblemId() {
     }
 
     @Override
-    public SolverStatus getSolverStatus() {
+    public synchronized SolverStatus getSolverStatus() {
         return solverStatus;
     }
 
     @Override
     public Solution_ call() {
-        solverStatus = SolverStatus.SOLVING_ACTIVE;
+        synchronized (this) {
+            if (solverStatus != SolverStatus.SOLVING_SCHEDULED) {
+                // This job has been canceled before it started
+                return problemFinder.apply(problemId);
+            }
+            solverStatus = SolverStatus.SOLVING_ACTIVE;
+        }
```

Comment:
@triceo I actually originally tried an AtomicRefence, for the reasons you mentioned.
Trouble was the code above versus the terminateEarly() code.
That latter did change drastically, so maybe there's a way to reintroduce the AtomicReference, let me try.